crates/sui-types/src/crypto.rs

```diff
@@ -33,21 +35,129 @@ use crate::error::{SuiError, SuiResult};
 use crate::sui_serde::{Base64, Readable, SuiBitmap};
 pub use enum_dispatch::enum_dispatch;
 
-// Comment the one you want to use
-
 // Authority Objects
 pub type AuthorityKeyPair = Ed25519KeyPair;
 pub type AuthorityPrivateKey = Ed25519PrivateKey;
 pub type AuthorityPublicKey = Ed25519PublicKey;
 pub type AuthoritySignature = Ed25519Signature;
 pub type AggregateAuthoritySignature = Ed25519AggregateSignature;
 
-// Account Objects
+// TODO(joyqvq): prefix these types with Default, DefaultAccountKeyPair etc
 pub type AccountKeyPair = Ed25519KeyPair;
 pub type AccountPublicKey = Ed25519PublicKey;
 pub type AccountPrivateKey = Ed25519PrivateKey;
 pub type AccountSignature = Ed25519Signature;
 
+#[enum_dispatch]
+#[derive(Debug)]
+pub enum SuiKeyPair {
+    Ed25519KeyPair,
+    Secp256k1KeyPair,
+}
+
+#[derive(Debug)]
+pub enum PublicKey<'a> {
+    Ed25519KeyPair(&'a Ed25519PublicKey),
+    Secp256k1KeyPair(&'a Secp256k1PublicKey),
+}
```

Comment:
In general, it's going to be a lot more complex to manage a struct that embarks references to public keys. Those keys are morally 32-ish bytes and [Copy]. It should be cheap to clone them. We should err on the side of simplicity and make the type embark a copy of each variant.